text/0000-module-resolution.md

```diff
@@ -0,0 +1,340 @@
+// importer: fancy/bar/bar.js
+import button from "ui/button"; // should resolve against ui/button@1.0.0
+```
+
+This means that from a developers perspective you just import the module you need.
+
+## State of the art
+
+An important topic to understand first is how module resolution works today in different technologies and host environments.
+
+JavaScript did not had built-in support for modules, but the community has created impressive work-arounds over the years. The two most important (and unfortunately incompatible) standards are: CommonJS (used fundamentally by Node - with some extra semantics) and AMD (Asyncronous module definition), used by RequireJs and a lot of other bundlers.
+
+Note that there are other non-standard formats, loaders and tools like UMD, SystemJS, IIFE, etc, which are interesting for you know and understand but are out of the scope for this proposal.
+
+### Node.js resolution
+
+If you ever used Node, you know that in order to import a module you need to use the `require()` method, and if you want to export something to some other module or file you must use `module.exports`. That is know nas [CommonJS format](https://en.wikipedia.org/wiki/CommonJS). Now if you want to import a 3party package how does that works? Well Node decided to implement its own resolution by looking at the content of a the node_modules folder based on some dependencies defined on `package.json`. You can look at the full algorithm [here](https://nodejs.org/docs/latest/api/modules.html#modules_all_together).
+
+All of this being said after years of very hard work, Node is finnaly addopting ES6 modules (which we will see in the next section) since LTS v12 and they are hoping to [fully transition to it soon](https://twitter.com/MylesBorins/status/1189618753065144322)
+
+Just to put all of this in a simple example:
+
+```js
+// This will be resolved by looking a node_modules/ folder "foo-bar" 
+// recursively upwards, and the check the package.kson to find the entry point. 
+const x = require('foo-bar') // This is using CommonJS syntax
+
+// This indicates that we are exporting a default function 
+// to whoever is requesting this module
+module.exports = function () { return x() }; // This is using CommonJS syntax
+
+```
+
+### ES6 Modules and browser resolution
+
+At the end of July 2014, TC39 (Javascript standarization comitee) had another meeting, during which the last details of the ECMAScript 6 (ES6) module syntax were finalized. [Here is the specification text](https://www.ecma-international.org/ecma-262/6.0/#sec-imports).
+
+The fundamental goal for ECMAScript 6 modules was to create a standard format that all users from the previous formats where comfortable and happy about, and lacked none of the caveats and issues - and since was part of the standar a lot of things could be improved like 1) The structure which can be statically analyzed (for static checking, optimizations, etc.) 2) Support for cyclic dependencies in a much better way.
```

Comment:
```suggestion
The fundamental goal for ECMAScript 6 modules was to create a standard format that all users from the previous formats were comfortable and happy about, and lacked none of the caveats and issues - and since was part of the standard a lot of things could be improved like

1. The structure which can be statically analyzed (for static checking, optimizations, etc.)
2. Better support for cyclic dependencies
```